### DISCUSSION TITLE
Usage with Storybook?

### DISCUSSION THREAD

**Discussion by MerlinMason** [TanStack/router#952] [Q&A]
Hi there,

We're currently using React-Router and considering migrating to Tanstack Router... 

We use Storybook a lot and we needed to use an [add-on](https://storybook.js.org/addons/storybook-addon-react-router-v6) to enable React Router to work with Storybook.

Is this also the case with Tanstack Router? and if so does such an add-on exist yet?

Thanks!

**Comment by erickarnis-tb** [comment]
Currently with Storybook I get this error on most stories: ReferenceError: Cannot access 'Root' before initialization
I'm on  @tanstack/react-router: "1.4.6",
Haven't been able to figure it out

**Reply by erickarnis-tb** [reply]
I figured it out. Importing anything from a router file leads to the execution of the router file which can lead to errors even in unrelated stories.

**Comment by dohomi** [comment]
I would like to know how to use Storybook as well. I think a plugin would be great, or simply a working solution how to create the router context for stories that you can display components which includes any API of the router

**Reply by Dixtir** [reply]
@dohomi have you found a solution for this?

**Reply by dohomi** [reply]
I use this:
```ts
const rootRoute = new RootRoute()
const indexRoute = new Route({ getParentRoute: () => rootRoute, path: "/" })
const memoryHistory = createMemoryHistory({ initialEntries: ["/"] })
const routeTree = rootRoute.addChildren([indexRoute])
const router = new Router({ routeTree, history: memoryHistory })

// @ts-ignore todo maybe soon a better solution for this?
export const withSbTanstackRouter: Preview["decorators"][0] = (Story, context) => {
  return <RouterProvider router={router} defaultComponent={() => <Story {...context} />} />
}
```
then wrap your most outer global decorator story like this:
```ts
export const SbDecorator: Decorator = (Story,args) => (
  <>{withSbTanstackRouter(Story,args))}</>
)
```

**Reply by Dixtir** [reply]
@dohomi thanks, much appreciated!

**Comment by alireza-mh** [comment]
I'm interested to know if there are any workaround/solutions to render the `Link` component inside a story. Some components might want to use router hooks even.

**Reply by Dixtir** [reply]
@alireza-mh have you found a solution for this?

**Comment by jaens** [comment]
Here's what I created: [Storybook fake Tanstack router - Gist](https://gist.github.com/jaens/693bd767b0f18d6577265815f6831c7a)

**Comment by boonya** [comment]
Here is what I've done

```ts
import type {PartialStoryFn, StoryContext} from '@storybook/types';
import {createMemoryHistory, createRootRoute, createRoute, createRouter, RouterProvider} from '@tanstack/react-router';

export default function withRouter(Story: PartialStoryFn, {parameters}: StoryContext) {
	const {initialEntries = ['/'], initialIndex, routes = ['/']} = parameters?.router || {};

	const rootRoute = createRootRoute();

	const children = routes.map((path) =>
		createRoute({
			path,
			getParentRoute: () => rootRoute,
			component: Story,
		}),
	);

	rootRoute.addChildren(children);

	const router = createRouter({
		history: createMemoryHistory({initialEntries, initialIndex}),
		routeTree: rootRoute,
	});

	return <RouterProvider router={router} />;
}

declare module '@storybook/types' {
	interface Parameters {
		router?: {
			initialEntries?: string[];
			initialIndex?: number;
			routes?: string[];
		};
	}
}
```

You can add it as global decorator in `preview.tsx` file.

**Reply by thienandangthanh** [reply]
@boonya Thank you for your snippet!
It works for my projects using @tanstack/react-router and Storybook.

I think your discussion should be `marked as answer` for this question.
And if possible, it should be also added to @tanstack/react-router docs.

**Reply by boonya** [reply]
Moreover, if you use `@storybook/addon-actions` you can integrate it as well.
```tsx
import {action} from '@storybook/addon-actions';

...
const children = routes.map((path) => createRoute({
  path,
  getParentRoute: () => rootRoute,
  component: Story,
  onLeave: action('router onLeave'),
  onEnter: action('router onEnter'),
  onStay: action('router onStay'),
})
...
```

**Reply by wurmr** [reply]
This worked great for me until router version 1.30. At that version and above it starts throwing a typescript error assigning `router={router}`.

**Reply by boonya** [reply]
@wurmr , I have `1.32.3` and it still works fine

**Reply by d4vidsha** [reply]
@boonya Can you explain further how to achieve this for newbies? I am one of them.

**Reply by boonya** [reply]
@d4vidsha what do you mean by "achieve this"? I have it working if it doesn't? I don't know. It depends on many factors - set of dependencies, storybook setup, browser and many others.

**Reply by d4vidsha** [reply]
@boonya Ah I was just wondering where the above code snippet should be placed more generally, as I'm a little lost even after reading up on the docs for [Storybook decorators](https://storybook.js.org/docs/writing-stories/decorators). Just trying to link your comment and those docs together in my head to create a minimum reproducible example.

**Reply by boonya** [reply]
I would recommend you remove as many as you can and leave the only code related to router. Then check how it work. If does, add other feature one by one to catch the guilty one.

**Comment by bhaugeea** [comment]
Here's something simple that's been working fine with router v1.4 and storybook v7.6. You can see it has no business logic at all, but it's enough to provide the router context so things don't break.

```tsx
// preview.tsx

import type { Preview } from '@storybook/react';
import { RouterProvider, createMemoryHistory, createRootRoute, createRouter } from '@tanstack/react-router';

const preview: Preview = {
  decorators: [
    (Story) => {
      return <RouterProvider router={createRouter({
        history: createMemoryHistory(),
        routeTree: createRootRoute({
          component: Story
        })
      })} />
    }
  ]
};

export default preview
```

**Reply by boonya** [reply]
Okay I see, but how about the situation you need to render your component by the specific route? How can I do that?

**Reply by bhaugeea** [reply]
Not sure, but in my opinion a component like that doesn't belong in storybook. I use storybook for things that are supposed to be generically composable. Buttons, modals, dropdowns, cards, etc. The only story I currently have that relies on router is a reusable implementation of tanstack table that puts its pagination, sorting, and filtering state in the URL. Doesn't care what route it's on, it just adds its own state.

If you need to recreate some non-trivial route tree, data fetching, and other stuff to make your stories work, then either: (a) you're trying to take storybook too far or (b) your components manage state badly and need to be written cleaner to be truly reusable.

**Reply by boonya** [reply]
"Too far" is too abstract statemnt. If you work with components library, you are correct. But what if you want to implement some component of real application which is dependent on router context as well as some data API. Does that mean I should not do so? But why? Storybook is amazing tool which gives me such ability throught it's decorators API and parameters.
So, what is "too far" to you may be "not enough" to someone other.
With that toolset storybook helps me implement frontend far ahead of backend in my projects. It save us a tone of time and make my team free of redundant headache.

**Comment by jaens** [comment]
Just FYI, there's quite a bit of subtlety involved with the Storybook (8) integration, the simple solution I and others have pasted above does not always work properly when switching between stories:
- it will keep a stale component (not re-rendered)
- once you navigate away from the default route, it will not navigate back

**Comment by levidrain** [comment]
I bypassed the RouterProvider and used the RouterContextProvider and so far it works well.  Can't promise it wouldn't break in a future version or that it works in all cases.  
In the `preview.ts` file add a decorator, like:

```
  const router = createRouter({ routeTree });
  ...
  decorators: [story => <RouterContextProvider router={router}>{story()}</RouterContextProvider>]
```

**Reply by jaens** [reply]
This will not work when the story actually navigates anywhere (eg. in more "system-level" stories / form submits).

(note that changing eg. hash parameters due to eg. radio group selection is also considered a navigation, but depending on the exact state management setup it might still work)

**Comment by dohomi** [comment]
I tried all the above ways but it all looses either strong typings on `useSearch` or `useParams` and the user is forced to relax the typings with `strict: false` which is not really ideal. I would like to display landing pages of tanstack router pages and if there is any business logic involved which is strongly typed I receive error messages. 

```
Invariant failed: Could not find a nearest match!
Invariant failed: Could not find an active match from "/_dashboard/team/$teamId/players"
```

I start to provide my own useSearch which basically keeps the same typings but replaces the options with {strict:false} 

```ts
import {
  type AnyRoute,
  type FullSearchSchema,
  type RegisteredRouter,
  type RouteById,
  type RouteIds,
  useSearch
} from "@tanstack/react-router"
import { isStorybook } from "../lib/generalConfig"

type StringLiteral<T> = T extends string ? (string extends T ? string : T) : never

type StrictOrFrom<TFrom, TStrict extends boolean = true> = TStrict extends false
  ? {
      from?: never
      strict: TStrict
    }
  : {
      from: StringLiteral<TFrom> | TFrom
      strict?: TStrict
    }

type UseSearchOptions<TFrom, TStrict extends boolean, TSearch, TSelected> = StrictOrFrom<
  TFrom,
  TStrict
> & {
  select?: (search: TSearch) => TSelected
}
type Expand<T> = T extends object
  ? T extends infer O
    ? O extends Function
      ? O
      : { [K in keyof O]: O[K] }
    : never
  : T

/**
 * Enhanced useSearch hook that optionally disables strict mode if running in Storybook.
 */
export function usePtSearch<
  TRouteTree extends AnyRoute = RegisteredRouter["routeTree"],
  TFrom extends RouteIds<TRouteTree> = RouteIds<TRouteTree>,
  TStrict extends boolean = true,
  TSearch = TStrict extends false
    ? FullSearchSchema<TRouteTree>
    : Expand<RouteById<TRouteTree, TFrom>["types"]["fullSearchSchema"]>,
  TSelected = TSearch
>(opts: UseSearchOptions<TFrom, TStrict, TSearch, TSelected>): TSelected {
  // Create a modified options object if in Storybook mode

  // Call the useSearch hook with the appropriate options
  return useSearch(isStorybook ? { strict: false } : opts)
}
```

**Reply by hiima** [reply]
@dohomi 
How is `../lib/generalConfig` implemented? I couldn't figure out how to detect if Storybook is running.

**Reply by dohomi** [reply]
its simply a `process.env.STORYBOOK` lookup. I set this to truthy

**Reply by hiima** [reply]
Got it, thanks!

**Reply by jmurga97** [reply]
I've been working on this for days. One thing that i found is how /_dashboard/team/$teamId/player is diferrent from /_dashboard/team/$teamId/player/. If this file is a index one on a file based routing the custom hooks inside wont recognize the actual route until is proper configured

**Comment by MoFarah9** [comment]
Here's what worked for me on a project with file-based routing. I simply create a router with a route for the story, and depending on the route I render a passthrough layout route if there's any passless (layout) on the route. It defaults to an index route for stories of components with no strict route APIs. I do on this component:
```tsx
export function StoryRouterProvider(
  props: PropsWithChildren<{
    parameters: StoryContext["parameters"];
  }>,
): ReactElement {
  const { children, parameters } = props;
  let storyRoutePath = "/";
  let layoutRouteId;

  // If the route parameter exists, extract any layout segment from the path,
  // if there's any layout segment, a layout route will be created as a parent of the story route
  if (parameters["route"]) {
    const pathSegments = parameters["route"].id.split("/").filter((e) => e);
    const layoutSegments = pathSegments.slice(
      0,
      pathSegments.findLastIndex((s) => s.startsWith("_")) + 1,
    );
    layoutRouteId = layoutSegments.join("/") || undefined;
    storyRoutePath = "/" + pathSegments.slice(layoutSegments.length).join("/");
  }

  const rootRoute = createRootRoute({});
  let layout =
    layoutRouteId &&
    createRoute({
      component: Outlet,
      id: layoutRouteId,
      getParentRoute: () => rootRoute,
    });
  const storyRoute = createRoute({
    component: () => children,
    path: storyRoutePath,
    getParentRoute: () => (layout ? layout : rootRoute),
  });
  layout?.addChildren([storyRoute]);
  rootRoute.addChildren([layout ? layout : storyRoute]);

  const finalPath = parameters["route"]?.params
    ? interpolatePath({
        path: storyRoutePath,
        params: parameters["route"].params,
      }).interpolatedPath
    : storyRoutePath;

  const router = createRouter({
    history: createMemoryHistory({
      initialEntries: [finalPath],
    }),
    routeTree: rootRoute,
  });

  return <RouterProvider router={router} defaultComponent={() => children} />;
}
```
Then I wrap the Story decorator on `preview.js` with the above component, then on the stories files, if a component uses route-specific APIs like  `getRouteApi`, I provide a `route`  parameter with the route ID and values for path parameters if there's any, for example
```ts
  parameters: {
    route: {
      id: "_authenticated/users/$userId/profile",
      params: { userId: 1 },
    },
  }
```

**Comment by jmurga97** [comment]
Hey guys, based on the previous answers i came up with a solution that supports the use of hooks, layouts and dynamic parameters.

I just wish to find a way to handle navigation flows based on the array entries and the generated route tree like this example with react router and mock service worker: https://github.com/mswjs/msw-storybook-addon/blob/main/packages/docs/src/demos/react-router-react-query/App.stories.tsx

```typescript
export interface RouteData {
  path: string;
  loader?: Record<string, unknown>;
}

interface InitialEntries {
  entry: string;
  params?: Record<string, string>;
}

interface RouterParameters {
  initialEntries?: InitialEntries[];
  routes?: RouteData[];
}

const defaultRoute = [{ path: '/' }];
const defaultInitialEntries = [{ entry: '/', params: undefined }];

export function withTanstackRouter(Story: PartialStoryFn, { parameters }: StoryContext) {
  const { initialEntries = defaultInitialEntries, routes = defaultRoute } = parameters?.router || {};
  const rootRoute = createRootRoute({});

  routes.forEach((route) => {
    const { parentRoute, storyRoute } = createRouteConfig(route, Story, rootRoute);

    if (parentRoute !== rootRoute) rootRoute.addChildren([parentRoute]);
    parentRoute.addChildren([storyRoute]);
  });

  // Process entries with dynamic parameters
  const processedEntries = initialEntries.map(({ entry, params }) => {
    if (params) {
      return interpolatePath({
        path: entry,
        params
      }).interpolatedPath;
    }
    return entry;
  });

  const router = createRouter({
    history: createMemoryHistory({
      initialEntries: processedEntries,
      initialIndex: 0
    }),
    routeTree: rootRoute
  });

  return <RouterProvider router={router}></RouterProvider>;
}

declare module '@storybook/types' {
  interface Parameters {
    router?: RouterParameters;
  }
}

/**
 * Creates route configuration for Storybook stories
 * @param {RouteData} route - Full story path pattern and loader if exists
 * @param {PartialStoryFn} Story - Storybook component
 * @param {RootRoute} rootRoute - Application root route
 * @returns {Object} Contains layout and story routes
 */

export function createRouteConfig(route: RouteData, Story: PartialStoryFn, rootRoute: RootRoute) {
  const { layoutRouteId, storyRoutePath } = parseRoutePath(route.path);

  const parentRoute = layoutRouteId
    ? createRoute({
        id: layoutRouteId,
        component: Outlet,
        getParentRoute: () => rootRoute
      })
    : rootRoute;

  const storyRoute = createRoute({
    component: Story,
    path: `${storyRoutePath}`,
    getParentRoute: () => parentRoute,
    loader: () => route.loader
  });

  return { parentRoute, storyRoute };
}

/**
 * Finds the index of the last layout segment in a route.
 * @param segments - Array of route segments.
 * @returns The index of the last layout found, or -1 if no layouts are present.
 */
const findLastLayoutIndex = (segments: string[]): number => {
  let lastLayoutIndex = -1;

  for (let i = segments.length - 1; i >= 0; i--) {
    if (segments[i].startsWith('_')) {
      lastLayoutIndex = i;
      break;
    }
  }
  return lastLayoutIndex;
};

/**
 * Parses a route and separates layout and story components.
 * @param path - Route to parse.
 * @returns Object with the layout route or undefined, and the story route path.
 * @throws Error if the path is invalid.
 */
export function parseRoutePath(path: string): RouteParseResult {
  if (!path || typeof path !== 'string') {
    throw new Error('Invalid path: must be a non-empty string');
  }
  const hasTrailingSlash = path.endsWith('/');

  const pathSegments = path.split('/').filter(Boolean);

  const layoutMarkerIndex = findLastLayoutIndex(pathSegments);
  const hasLayout = layoutMarkerIndex !== -1;

  const boundaryIndex = layoutMarkerIndex + 1;
  const layoutSegments = hasLayout ? pathSegments.slice(0, boundaryIndex) : [];
  const storySegments = hasLayout ? pathSegments.slice(boundaryIndex) : [...pathSegments];

  const layoutRouteId = layoutSegments.length ? `/${layoutSegments.join('/')}` : undefined;
  let storyRoutePath = storySegments.length ? `/${storySegments.join('/')}` : '/';

  if (hasTrailingSlash && storyRoutePath !== '/') storyRoutePath += '/';

  return {
    layoutRouteId,
    storyRoutePath
  };
}
```
# Usage

### Basic Route

```tsx
export const Basic = {
  parameters: {
    router: {
      routes: [{ path: '/dashboard' }],
      initialEntries: [{ entry: '/dashboard' }]
    }
  }
};
```

### Route with Dynamic Parameters

```tsx
export const WithParams = {
  parameters: {
    router: {
      routes: [{ path: '/users/:id' }],
      initialEntries: [
        {
          entry: '/users/:id',
          params: { id: '123' }
        }
      ]
    }
  }
};
```

### Route with Dynamic Parameters and loader

```tsx
export const WithParams = {
  parameters: {
    router: {
      routes: [
        { 
          path: '/_layout/users/:id',
          loader: {
            user: {id: '123', name:'Lorem Ipsum'}
          }
        }
      ],
      initialEntries: [
        {
          entry: '/users/:id',
          params: { id: '123' }
        }
      ]
    }
  }
};
```

**Reply by EskiMojo14** [reply]
Nice! By the way, you can improve the type safety of the types using the `FileRoutesByFullPath` type exported from the routeTree.gen file:

```ts
// take an intersection and combine it for better hovers
export type Compute<T> = { [K in keyof T]: T[K] } & unknown;

export type IfMaybeUndefined<T, True, False> = undefined extends T
  ? True
  : False;

export type RequiredKeys<T> = {
  [K in keyof T]: undefined extends T[K] ? never : K;
}[keyof T];

export type HasRequiredKeys<T, True, False> =
  RequiredKeys<T> extends never ? False : True;

export type RouteData = {
  [Path in keyof FileRoutesByFullPath]: Compute<
    { path: Path } & IfMaybeUndefined<
      // require loader data if we need it (i.e. it can't be undefined)
      FileRoutesByFullPath[Path]["types"]["loaderData"],
      {
        loaderData?: FileRoutesByFullPath[Path]["types"]["loaderData"];
      },
      {
        loaderData: FileRoutesByFullPath[Path]["types"]["loaderData"];
      }
    >
  >;
}[keyof FileRoutesByFullPath];

export type InitialEntries = {
  [Path in keyof FileRoutesByFullPath]: Compute<
    { entry: Path } & HasRequiredKeys<
      // require params if we have any (no need to provide {})
      FileRoutesByFullPath[Path]["types"]["params"],
      {
        params: FileRoutesByFullPath[Path]["types"]["params"];
      },
      {
        params?: FileRoutesByFullPath[Path]["types"]["params"];
      }
    >
  >;
}[keyof FileRoutesByFullPath];
```

This means you'll get proper type checking for loader data and params (and errors if they're wrong or missing).

**Comment by llite22** [comment]
```javascript
import { Decorator } from '@storybook/react';
import {
createRootRoute,
createRouter,
RouterProvider
} from '@tanstack/react-router';

export const RouterDecorator: Decorator = (Story) => {
const rootRoute = createRootRoute({
component: () => <Story />
});

const routeTree = rootRoute;

const router = createRouter({
routeTree
});

return <RouterProvider router={router} />;
};
```

**Reply by Shuunen** [reply]
For those who want to check the whole file, here's my `.storybook/preview.tsx` file based on @llite22 suggestion : 

```tsx
import type { Decorator, Preview } from "@storybook/react";
import { createRootRoute, createRouter, RouterProvider } from "@tanstack/react-router";
import React from "react";
import "../src/index.css";

const RouterDecorator: Decorator = (Story) => {
  const rootRoute = createRootRoute({ component: () => <Story /> });
  const routeTree = rootRoute;
  const router = createRouter({ routeTree });
  return <RouterProvider router={router} />;
};

const preview: Preview = {
  decorators: [RouterDecorator],
};

export default preview;
```

Works great thanks <3

**Reply by devDanielCespedes** [reply]
> ```js
> import { Decorator } from '@storybook/react';
> import {
> createRootRoute,
> createRouter,
> RouterProvider
> } from '@tanstack/react-router';
> 
> export const RouterDecorator: Decorator = (Story) => {
> const rootRoute = createRootRoute({
> component: () => <Story />
> });
> 
> const routeTree = rootRoute;
> 
> const router = createRouter({
> routeTree
> });
> 
> return <RouterProvider router={router} />;
> };
> ```

I love you <3

**Reply by theoludwig** [reply]
Work great, until you use `import { getRouteApi } from "@tanstack/react-router"` in your component.

Any way to make Storybook work when using hooks like `Route.useSearch()`?

**Reply by jaens** [reply]
@theoludwig Read the fine discussion here, solution is already mentioned...

**Reply by theoludwig** [reply]
Other solutions will display the whole route/page. No solution here allows to show only 1 component in isolation while still having everything TanStack Start/Router work, or I missed it, please link me to it, thanks!

**Comment by jasikpark** [comment]
I just decided to load the entire routeTree 🤷 

```tsx
import type { Meta, StoryObj } from '@storybook/react-vite';
import { createMemoryHistory, createRouter, RouterProvider } from '@tanstack/react-router';
import { expect, userEvent, within } from 'storybook/test';
import { routeTree } from '../../routeTree.gen';

type Args = {
  initialEntries: string[];
  initialIndex: number;
};

function Routes({ initialEntries, initialIndex }: Args) {
  const router = createRouter({
    history: createMemoryHistory({ initialEntries, initialIndex }),
    routeTree,
  });

  return <RouterProvider router={router} />;
}

// More on how to set up stories at: https://storybook.js.org/docs/writing-stories#default-export
const meta = {
  title: 'Routes/Enroll',
  component: Routes,
  args: {
    initialIndex: 0,
  },
} satisfies Meta<typeof Routes>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Success: Story = {
  args: {
    initialEntries: ['/enroll'],
  },
  async play({ canvasElement }) {
    const { findByRole, findByText } = within(canvasElement);
    await userEvent.type(await findByRole('textbox', { name: 'Enrollment code' }), '1234{enter}');
    await expect(await findByText('Computer successfully enrolled:')).toBeInTheDocument();
  },
};
```

Definitely is a bit more integration testing-ish, but I can just set the proper intitialEntries and go from there.

**Comment by pedaars** [comment]
We managed to get hooks working like so 

```
//decorator
import type { Decorator } from '@storybook/react'
import {
  createRootRoute,
  createRouter,
  RouterProvider,
  createRoute,
  createHashHistory,
} from '@tanstack/react-router'

export const RouterDecorator: Decorator = (Story) => {
  const Root = createRootRoute({ component: () => <Story /> })

  const HomeRoute = createRoute({
    getParentRoute: () => Root,
    path: '/',
  })

  const routeTree = Root.addChildren([HomeRoute, TestRoute])
  const hashHistory = createHashHistory()

  const router = createRouter({
    routeTree,
    history: hashHistory,
  })

  return <RouterProvider router={router} />
}
```


Then in our story which showcases a navigation blocker we can use the useNavigation hook to navigate({ to: '/' }) and trigger the Block component

**Comment by MatejBransky** [comment]
I put together this solution (~draft) that addresses my main issue: displaying a full page with type-safe specific search or path parameters in Storybook using TanStack Router. I managed to get it working with this `<WithRouter />` component. I haven't extensively tested it yet, but it seems to have some potential for others facing similar challenges.

```tsx
import {
  createMemoryHistory,
  createRootRoute,
  createRoute,
  createRouter,
  Outlet,
  RouteIds,
  RouterProvider,
  NavigateOptions,
  RouteById,
} from '@tanstack/react-router'
import { TanStackRouterDevtools } from '@tanstack/react-router-devtools'

import { router } from '@shared/router'

export type ExtractRouteOptions<TId extends RouteIds<(typeof router)['routeTree']>> = Omit<
  NavigateOptions<typeof router, string, TId>,
  'to' | 'search' | 'params'
> & {
  search?: Partial<RouteById<(typeof router)['routeTree'], TId>['types']['fullSearchSchema']>
} & (RouteById<(typeof router)['routeTree'], TId>['types']['allParams'] extends Record<
    string,
    never
  >
    ? { params?: undefined }
    : { params: RouteById<(typeof router)['routeTree'], TId>['types']['allParams'] })

/**
 * A utility component to wrap children with a mocked router for testing purposes.
 * It allows you to specify the initial route, search parameters, and route parameters.
 *
 * @param children - The child components to be rendered within the router context.
 * @param routeId - The ID of the route to navigate to initially.
 * @param search - Optional search parameters for the route.
 * @param params - Optional route parameters.
 *
 * @returns A RouterProvider component that provides the mocked router context.
 */
export const WithRouter = <
  TRouter extends typeof router,
  const TId extends RouteIds<TRouter['routeTree']>,
>({
  children,
  routeId,
  search,
  params,
}: React.PropsWithChildren<
  {
    routeId: TId
  } & ExtractRouteOptions<TId>
>) => {
  const history = createMemoryHistory()
  const rootRoute = createRootRoute({
    component: () => (
      <>
        <Outlet />
        <TanStackRouterDevtools />
      </>
    ),
  })
  const mockedRouter = createRouter({
    history,
    routeTree: rootRoute.addChildren([
      createRoute({
        getParentRoute: () => rootRoute,
        path: routeId,
        component: () => children,
      }),
    ]),
  })

  mockedRouter.navigate({ to: routeId.replace(/\/$/, ''), search, params } as any)

  return <RouterProvider router={mockedRouter} />
}
```

The usage looks like this:

```tsx
import type { Meta, StoryObj } from '@storybook/react'
import { WithRouter } from '@shared/lib/test'
import { YourPageComponent } from './YourPageComponent'

const meta: Meta<typeof YourPageComponent> = {
  component: YourPageComponent,
}
export default meta
type Story = StoryObj<typeof YourPageComponent>

export const Example: Story = {
  render: () => (
    <WithRouter routeId="/your-route/$entityId" params={{ entityId: 'abc' }}>
      <TransformationPlansListingPage />
    </WithRouter>
  ),
}
```

**Reply by cvoege** [reply]
Thank you @MerlinMason! This works for my purposes! created a decorator based off of this, since I find that format slightly less fragile so you don't deal with stuff like passing down props in `render`, if anyone is interested:


```tsx
export const withTanstackRouter = <
  TRouter extends typeof router,
  const TId extends RouteIds<TRouter["routeTree"]>
>(options: {
    routeId: TId
  } & ExtractRouteOptions<TId>) => {
  return (Story: React.ComponentType): JSX.Element => {
    return (
      // @ts-expect-error Couldn't get this to internally type nicely, but it works and the external facing types are accurate.
      <WithRouter<TRouter, TId> {...options}>
        <Story />
      </WithRouter>
    );
  };
};

```

Usage:
```tsx
const meta: Meta<typeof Link> = {
  title: "UI/Navigation/Link",
  component: Link,
  decorators: [
    withTanstackRouter({ routeId: "/_auth/$orgName/system/", params: { orgName: "example" } }),
  ],
};
```

**Reply by bavjackson** [reply]
I've been trying both the component and decorator with success so far. The only thing I've changed was mocking out the `navigate` function after the initial navigate to the route.
```typescript
// At the top of your file
import { fn } from "storybook/test";
/*
.... the rest of the component here
*/
  mockedRouter.navigate({
    to: routeId.replace(/\/$/, ""),
    search,
    params,
  } as any);
  mockedRouter.navigate = fn();  <<< Mock navigate here

  return <RouterProvider router={mockedRouter} />;
```
This means that a few components with links in I want to use won't just 404 if someone clicks them accidentally. I didn't find a clean way of redirecting the various routes back to the one I had rendered in the mock router, and I didn't want it to distract from the component being tested.

**Comment by hollandjake** [comment]
they've now released @storybook/tanstack-react to bring official support to storybook.

https://github.com/storybookjs/storybook/issues/34284 
https://storybook.js.org/docs/get-started/frameworks/tanstack-react

**Reply by EskiMojo14** [reply]
this is really nice

the only potentially strange behaviour i've found is that `createIsomorphicFn`'s mocking seems to be order dependent - whichever implementation is provided *last* decides whether it's treated as a client function with an implementation or a server function that should throw

```ts
const hello = createIsomorphicFn()
  .client(() => console.log("hi from client"))
  .server(() => console.log("hi from server"));
  
// becomes
const hello = fn();

// whereas

const hello = createIsomorphicFn()
  .server(() => console.log("hi from server"))
  .client(() => console.log("hi from client"));
  
// becomes
const hello = fn(() => console.log("hi from client"));
```

this makes *some* sense, but it'd be potentially nicer to just look for a client implementation and use it if available, otherwise no-op

**Reply by hollandjake** [reply]
@EskiMojo14 yeah ive been seeing a lot of issues with it, as ive been trying to use it but its just too many issues for me atm. im currently using `storybook-addon-tanstack-start` which seems to be much better at not screwing with your files as part of a vite plugin. however it does not support route files atm

**Comment by hollandjake** [comment]
@EskiMojo14 incase this helps you, i just managed to make storybook work without code transformation and it can run route files with full navigation support.

tutorial ang guide here.

https://gist.github.com/hollandjake/6a28d71b6cc24a2714f9b04f3ba06ff4

**Comment by chrisjcthomas** [comment]
You usually do not need a Storybook addon just to make TanStack Router APIs available. A Storybook decorator that creates a memory-history router is enough for many component stories.

The important detail is whether the component uses generic router APIs (`Link`, `useNavigate`) or route-specific APIs (`getRouteApi`, strict `useParams`, strict `useSearch`).

For generic components, a simple decorator works:

```tsx
import { RouterProvider, createMemoryHistory, createRootRoute, createRoute, createRouter } from '@tanstack/react-router'

export const withRouter = (Story, context) => {
  const path = context.parameters.router?.path ?? '/'

  const rootRoute = createRootRoute()
  const storyRoute = createRoute({
    getParentRoute: () => rootRoute,
    path,
    component: () => <Story />,
  })

  rootRoute.addChildren([storyRoute])

  const router = createRouter({
    routeTree: rootRoute,
    history: createMemoryHistory({ initialEntries: [path] }),
  })

  return <RouterProvider router={router} />
}
```

Then in `.storybook/preview.tsx`:

```tsx
export const decorators = [withRouter]
```

For route-specific components, the fake route must match the route the component expects. For example, if a component calls:

```ts
const route = getRouteApi('/_dashboard/team/$teamId/players')
```

then the Storybook router needs to include that same route id/path and initial params/search. Otherwise strict APIs correctly fail because there is no active matching route.

So the rule of thumb is:

- shared UI components: wrap with a simple memory router
- page/route components: build a Storybook route that mirrors the real route id/path/params
- avoid reusing one singleton router across all stories; create it per story so navigation state does not leak between stories

That is also why some older snippets work for `Link` but break for strict `useSearch` / `useParams`: they provide router context, but not the specific active route match the component is asking for.
